b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+                    'source_has_legal_hold':
+                        source_file_version.legal_hold is not LegalHold.UNSET,
+                }
+            )
+
+        if destination_file:
+            params.update(
+                {
+                    'destination_replication_status':
+                        destination_file.selected_version.replication_status,
+                }
+            )
+
+        if source_file and destination_file:
+            params.update(
+                {
+                    'metadata_differs':
+                        source_file.selected_version.file_info !=
+                        destination_file.selected_version.file_info,
+                }
+            )
+
+        return cls(**params)
+
+
+@dataclass
+class AbstractReplicationReport(metaclass=ABCMeta):
+    """
+    Aggregation of valuable information about file replication
+    after scanning source and (optionally) destination folders.
+    """
+
+    @abstractmethod
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path]):
+        pass
+
+
+@dataclass
+class CountAndSampleReplicationReport(AbstractReplicationReport):
+    """
+    Replication report which groups and counts files by their `ReplicationAttrs` and
+    also stores first and last seen examples of such files.
+    """
+    counter_by_status: Counter[ReplicationAttrs] = field(default_factory=Counter)
+    samples_by_status_first: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+    samples_by_status_last: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path] = None):
+        status = ReplicationAttrs.from_files(source_file, destination_file)
+        self.counter_by_status[status] += 1
+
+        sample = (
+            source_file and source_file.selected_version,
+            destination_file and destination_file.selected_version,
+        )
+        self.samples_by_status_first.setdefault(status, sample)
+        self.samples_by_status_last[status] = sample
+
+
+@dataclass
+class ReplicationMonitor:
+    """
+    Calculates source and (optionally) destination replication statistics.
+
+    :param b2sdk.v2.Bucket bucket: replication source bucket
+    :param b2sdk.v2.ReplicationRule rule: replication rule to be monitored;
+    should belong to `bucket`'s replication configuration
+    :param b2sdk.v2.B2Api destination_api: B2Api instance for destination
+    bucket; if destination bucket is on the same account as source bucket,
+    omit this parameter and then source bucket's B2Api will be used
+    :param type replication_report_class: subclass of ReplicationReport,
+    used to aggregate files to statistics
+    :param b2sdk.v2.Report report: instance of Report which will report
```

Comment:
it's a progress reporter, right? And we already have one implemented for uploads, downloads and sync

Reply:
Yeah, it was initially in `sync` module, but was extracted and now is used by `sync` and `scan`.